solidity/test/TokenTimeWeightedAverage.js

```diff
@@ -0,0 +1,373 @@
+                    twa.addPastSample(ZERO_ADDRESS, n, d, past, { from: seeder }),
+                    'ERR_INVALID_ADDRESS'
+                );
+            });
+
+            it('should revert when a seeder attempts to add a sample with a zero denominator', async () => {
+                const past = now.sub(new BN(1));
+
+                const n = new BN(1000);
+                const d = new BN(0);
+                await expectRevert(twa.addPastSample(token, n, d, past, { from: seeder }), 'ERR_ZERO_VALUE');
+            });
+
+            it('should revert when a seeder attempts to add a future sample', async () => {
+                const future = now.add(new BN(100));
+
+                const n = new BN(1000);
+                const d = new BN(2000);
+                await expectRevert(twa.addPastSample(token, n, d, future, { from: seeder }), 'ERR_INVALID_TIME');
+            });
+        });
+    });
+
+    describe('accumulating TWA', () => {
+        const expectAlmostEqual = (amount1, amount2, maxError = Decimal(0.00000001)) => {
+            if (!amount1.eq(amount2)) {
+                const error = Decimal(amount1.toString()).div(amount2.toString()).sub(1).abs();
+                expect(error.lte(maxError)).to.be.true(`error = ${error.toFixed(maxError.length)}`);
+            }
+        };
+
+        const token = accounts[8];
+        let acc;
+
+        const test = async (valueStepFunction, timeStepFunction) => {
+            let value = { n: new BN(0), d: new BN(0) };
+            let time = now;
+            const time0 = now;
+
+            for (let i = 1; i < 100; ++i) {
+                time = timeStepFunction(time);
+                await twa.setTime(time);
+
+                value = valueStepFunction(value);
+
+                await twa.addSample(token, value.n, value.d, { from: owner });
+                addTWASample(acc, value.n, value.d, time);
+
+                const twaValue = await twa.timeWeightedAverage.call(token, time0);
+                const testTwaValue = getTWA(acc, time0, time);
+                const twaDecValue = Decimal(twaValue[0].toString()).div(Decimal(twaValue[1].toString()));
+                expectAlmostEqual(twaDecValue, testTwaValue);
+            }
+        };
+
+        beforeEach(async () => {
+            await twa.initialize(token, now);
+            acc = initTWA(now);
+        });
+
+        it('should properly accumulate large values', async () => {
+            await test(
+                (value) => ({
+                    n: value.n.add(new BN(10 ** 10).mul(new BN(10).pow(new BN(18)))),
+                    d: new BN(1)
+                }),
+                (time) => time.add(duration.days(1))
+            );
+        });
+
+        it('should properly accumulate infinitesimal values', async () => {
+            await test(
+                (value) => ({
+                    n: new BN(1),
+                    d: value.d.add(new BN(10 ** 10).mul(new BN(10).pow(new BN(18))))
```

Comment:
Perhaps just `value.d.add(new BN(10).pow(new BN(28))))`?

Reply:
I wanted to be very explicit about the meaning of the numbers